### ISSUE TITLE
Two observed stepper signal discrepancies during diagonal moves

### ISSUE THREAD

**Issue by boelle** [MarlinFirmware/MarlinDev#35]
Continuation of https://github.com/MarlinFirmware/Marlin/issues/2370

**Comment by maicrodrop** [comment]
Small update:  Resetting TCNT1 to 0, if it's already passed OCR1A, works so far; been printing and testing with it.  Have also tested setting TCNT1 to (0.75*OCR1A) which works as well, though haven't tested it as thoroughly yet; this would be preferable, as it minimizes the dead time between ISR calls.

I've also been investigating the 1.5.6>1.5.7 compiler change further.  I added code to the ISR, to use the fastio.h code to write a pin HIGH at the beginning and LOW at the end of the ISR.  While this does change the absolute execution time of the ISR, compiling the same changed code in 1.5.6 and 1.5.7 at least gives a (theoretical) good relative comparison.  I found that an individual STEP pulse takes ~3.5µS in 1.5.5, and ~4.5µS in 1.5.7.  The entire ISR (measured between when the "ISR_signal" pin [defined in pins.h, so WRITE is used in the normal way] goes HIGH and then when it goes LOW) takes ~86µS in 1.5.5, and ~108µS in 1.5.7.  Still doesn't narrow down _why_, but at least shows that there is a definite execution time difference.

The four images below show the 'scope shots listed above; there are four-step-set captures compiled in 1.5.5 and 1.5.7, and single step zoom in both as well.  Yellow is Channel 1/X STEP; Blue is Channel2/Y STEP; Purple is Channel3/"ISR_signal", in these shots HIGH while the ISR is running (set HIGH when it starts, then set LOW at the very end).
![155_single](https://cloud.githubusercontent.com/assets/13168739/9093492/71747e18-3b7a-11e5-9975-4b16731f60ca.png)
![157_full](https://cloud.githubusercontent.com/assets/13168739/9093493/71833fb6-3b7a-11e5-9bbd-66391fe39d91.png)
![157_single](https://cloud.githubusercontent.com/assets/13168739/9093495/718c55d8-3b7a-11e5-9333-c06df7565edd.png)
![155_full](https://cloud.githubusercontent.com/assets/13168739/9093494/718a821c-3b7a-11e5-8bda-6ee181706081.png)


One final note on testing; I did also experiment with having the stepper interrupt write the timer and compare values to the serial port if TCNT1 ran past OCR1A; I saw quite a few instances of the timer being at 212, when it should have triggered the interrupt when it was 203, showing that the overrun is pretty small (I believe those numbers pretty closely match µS, so the ISR is overrunning when it should get triggered again by ~9µS, in those instances).  Again, this test does effect the ISR execution time, but at least does provide _some_ insight.

As always, let me know if you'd like more information, if I wasn't clear somewhere, or if you notice any egregious logic errors.

**Comment by thinkyhead** [comment]
Interesting results! I wonder if the compile flags are different between 1.5.5 and 1.5.7. Maybe the better performance of 1.5.5 can be restored by using a different optimization level. Or, perhaps this is due to library changes that we can't affect.

**Comment by Wackerbarth** [comment]
@thinkyhead -- Actually, if we go to IDE 1.6 (current), we CAN control things and use it to build from the core/ library/ compiler of our choice.

The Arduino IDE now includes a library distribution system to automatically offer upgraded components.
By specifying our own boards and cores, and even downloading archives of the older compilers, we can selectively mix and match each of the parts.

**Comment by oysteinkrog** [comment]
This sounds similar to an issue I had last year:
https://plus.google.com/+%C3%98ysteinKrog/posts/FPpTKnrfE27
The reason I kept scrolling up/down in the menus is that I suspected that the LCD cpu load would help trigger it.

**Comment by guyzmo** [comment]
Hi, 

for the note I went into the same issue on latest marlin with a board running at 1/16th microstepping (657 steps/mm for E drive).

I found out the extruder drive was stuttering starting at a feedrate of 913mm/s, and I could notice it as well on high speed X/Y moves, but I did not take data for those.

The gap is consistent being 32ms for every feedrate:

![accel_issue_g0_e-5_f1000-gap](https://cloud.githubusercontent.com/assets/254441/9381722/d60aae84-4731-11e5-8f20-ab2feda60cca.png)

And the gap is happening when when the pulsing period is reaching 200µs:

![accel_issue_g0_e-5_f1000-pulse_rate](https://cloud.githubusercontent.com/assets/254441/9381723/d981755c-4731-11e5-9d7d-ba9666a856fd.png)

I have added a global that takes the minimum calculated timer value, and when the gap is happening, I had a timer value of 202.

Because of the 32ms period we felt it was likely to be a race condition between `OCR1A` and `OVF`, as 1/2000000*65535=32ms. So a google search on *OCR1A race condition* led to the makergear forum post and then here.

The `if (TCNT1 >= OCR1A) TCNT1 = 0;` at the end of the `ISR` works great (with a negligeable delay visible on the scope), though I wonder if there wouldn't be room of optimization to avoid running into this issue.

HTH

**Comment by Wurstnase** [comment]
I've done something similar in my Due-Fork.

```
uint16_t counter = TCNT1 + 10; // 10 or whatever you want for a safe spot
if (counter >= OCR1A) OCR1A = counter;
```

**Comment by Wurstnase** [comment]
Look like that ```OCR1A = (OCR1A < (TCNT1 +16)) ? (TCNT1 + 16) : OCR1A;``` solve this issue.
http://forums.reprap.org/read.php?397,542552,549268#msg-549268

**Comment by guyzmo** [comment]
@Wurstnase do you have any rationale behind the `TCNT+16`? 
And will we catch it faster when that happens, than with the `TCNT1=0` fix?

I'm not really convinced this hack is a better one than the previous fix…

**Comment by Wurstnase** [comment]
> @Wurstnase do you have any rationale behind the TCNT+16? 
Not really. 8 μs just looks good for it.
>And will we catch it faster when that happens, than with the TCNT1=0 fix?
Yes. With TCNT=0 you reset it and it starts again.

**Comment by thinkyhead** [comment]
> 8μs just looks good for it.
@Wurstnase Naive question: Is it possible to use `F_CPU` in some way instead of `16` explicitly?

**Comment by Wurstnase** [comment]
I think we need fixed ticks. Not time.

**Comment by thinkyhead** [comment]
@Wurstnase What is meant by "fixed ticks"? I could use a primer to understand what is being done in your solution, and even to understand what `OCR1A` and `TCNT1` mean. Can you explain to me as if I were a small child? (Because in this matter, I truly am…)

**Comment by Wurstnase** [comment]
https://github.com/MarlinFirmware/Marlin/issues/2525#issuecomment-136880938

**Comment by thinkyhead** [comment]
Haha, thanks! I just saw that other comment too. :tongue:

**Comment by guyzmo** [comment]
After trying out both fixes, I pushed the `E` axis to `M203 E999999` and tried to see how it goes as I'm going stupidly fast. As a reminder I used to have an issue at around `F913`, and I used to have the issue double at around `F2200`.

Now it's all fine at `F2000`, `F3000`… up until reaching `F6000` where I had the following issue:

![G0_E5_F6000-2](https://cloud.githubusercontent.com/assets/254441/9793282/a1a7a668-57dc-11e5-9a2e-2e693fcf8835.png)

So it's doing a mini pause between each 4 pulses. Sounds like the original problem, so I did zoom in:

![G0_E5_F6000](https://cloud.githubusercontent.com/assets/254441/9793863/a1beb3fa-57df-11e5-8c9b-ab7928c72fe9.png)

and the gap is consistently 24ms. I then looked at when the issue starts, and it is at `F5980` it misbehaves at higher speeds, like the following at `F10000`:

![G0_E5_F10000](https://cloud.githubusercontent.com/assets/254441/9794353/8d301d72-57e2-11e5-88dc-436b0c97c079.png)

Obviously, it's less an issue than the original one — as it's unlikely people is going to drive extruders at that speed — but it still feels somehow related, even though it's more likely to be an issue with the way `calc_timer()` is doing its job. What concerns me is that it's not really the kind of issues I was expecting by running the steppers way too fast (I was more expecting irregular stepping).

I'm having a look at that, and I'll try to find a proper solution, but I'd be happy to hear your thoughts about it. I might open a new issue on the topic if you believe both issues are not related.

BTW, my steps/mm are `657` on the extruder.
